### PR TITLE
Made snapshot rebuild logic take target into account

### LINKED ISSUE #11400: `flutter run lib/foo.dart` always runs the first thing built
## Steps to Reproduce

1) Create a sample using `flutter create`
1) Rename `lib/main.dart` to `lib/blue.dart`
1) Copy `lib/blue.dart` to `lib/red.dart`
1) Edit `lib/red.dart` so that the theme color is red.
1) Execute `flutter run lib/blue.dart`
1) Notice that the theme is blue.
1) Execute `flutter run lib/red.dart`
1) Notice that the theme is _still_ blue.
1) Execute `flutter build clean`
1) Execute `flutter run lib/red.dart`
1) Notice that the theme is now red.

Basically, once you build an app, no amount of changing the arguments will let you run a different program until you do "flutter build clean" (or remove the build/ directory).

## Logs

```
Launching lib/red.dart on Pixel in debug mode...
Initializing gradle...                                0.6s
Resolving dependencies...                             0.9s
Running 'gradlew assembleDebug'...                    3.2s
Built build/app/outputs/apk/app-debug.apk (21.2MB).
Installing build/app/outputs/apk/app.apk...           7.2s
Syncing files to device Pixel...                      2.9s

🔥  To hot reload your app on the fly, press "r". To restart the app entirely, press "R".
An Observatory debugger and profiler on Pixel is available at: http://127.0.0.1:8112/
For a more detailed help message, press "h". To quit, press "q".

Application finished.
```

```
Analyzing /Users/gspencer/code/run_bug/lib/red.dart...
No issues found!
Ran in 3.9s
```

## Flutter Doctor

[✓] Flutter (on Mac OS X 10.12.6 16G29, locale en-US, channel master)
    • Flutter at /Users/gspencer/code/flutter
    • Framework revision 41d5ad411d (4 minutes ago), 2017-07-25 15:59:41 -0700
    • Engine revision 3a12bc092d
    • Tools Dart version 1.25.0-dev.7.0

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /Users/gspencer/Library/Android/sdk
    • Platform android-25, build-tools 25.0.3
    • ANDROID_HOME = /Users/gspencer/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 8.3.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.3, Build version 8E3004b
    • ios-deploy 1.9.1
    • CocoaPods version 1.2.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Ultimate Edition (version 2017.2)
    • Flutter plugin version 15.2
    • Dart plugin version 172.3317.48

[✓] Connected devices
    • Pixel • FA6AS0303682 • android-arm • Android 7.1.2 (API 25)

### PR DESCRIPTION
**Pull request by mravn-google**
Fixes #11400

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/build.dart [hunk 1]
```diff
@@ -8,37 +8,52 @@ import 'package:crypto/crypto.dart' show md5;
...
+/// This class can be used during build actions to compute the fingerprint of
+/// the build action inputs, and if unchanged from the previous build, skip the
 /// build step. This assumes that build outputs are strictly a product of the
-/// input files.
-class Checksum {
-  Checksum.fromFiles(Set<String> inputPaths) : _checksums = <String, String>{} {
-    final Iterable<File> files = inputPaths.map(fs.file);
+/// input files and properties provided.
+class Fingerprint {
+  Fingerprint.fromInputs({Set<String> filePaths, Map<String, String> properties}) : _fingerprint = <String, String>{} {
```
Comment (jakobr-google) [inline packages/flutter_tools/lib/src/base/build.dart]: Nit: `filePaths` could just be an `Iterable<String>`, right? Then you wouldn't need to convert the lists to sets in the tests.

File: packages/flutter_tools/test/base/build_test.dart [hunk 2]
```diff
@@ -20,48 +20,75 @@ void main() {
 
       testUsingContext('throws if any input file does not exist', () async {
         await fs.file('a.dart').create();
-        expect(() => new Checksum.fromFiles(<String>['a.dart', 'b.dart'].toSet()), throwsA(anything));
+        expect(() => new Fingerprint.fromInputs(filePaths: <String>['a.dart', 'b.dart'].toSet()), throwsA(anything));
```
Comment (jakobr-google) [inline packages/flutter_tools/test/base/build_test.dart]: `throwsArgumentError` instead of `throwsA(anything)`?

**Review by jakobr-google — APPROVED** [review]
LGTM

File: packages/flutter_tools/test/commands/build_flx_test.dart [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      temp = fs.systemTempDirectory.createTempSync('flutter_build_flx');
+      projectDir = temp.childDirectory('flutter_project');
+    });
+
+    tearDown(() {
+      temp.deleteSync(recursive: true);
+    });
+
+    testUsingContext('rebuilds if target changes', () async {
+      await runAsync(<String>['flutter', 'create', projectDir.path], allowReentrantFlutter: true);
```
Comment (jakobr-google) [inline packages/flutter_tools/test/commands/build_flx_test.dart]: This'll only work if `flutter` is on the path (and maybe not on Windows). The other tests use `createTestCommandRunner()`, maybe that instead?

File: packages/flutter_tools/lib/src/base/build.dart (hunk 1 above)
Comment (mravn-google) [inline packages/flutter_tools/lib/src/base/build.dart]: Done.

File: packages/flutter_tools/test/base/build_test.dart (hunk 2 above)
Comment (mravn-google) [inline packages/flutter_tools/test/base/build_test.dart]: Done.

**Comment by mravn-google** [comment]
This appeared to have already been done by #11820, but that PR only takes change of main path into account, if the different main scripts are not part of the app for any other reason than being the main entry point. Specifically, if you have `lib/blue.dart` and `lib/red.dart` with blue and red theme colors, respectively, and _which import each other_, then running first `flutter run lib/blue.dart` and then `flutter run lib/red.dart` does not change from blue to red.

**Comment by Hixie** [comment]
@cbracken will surely want to review this since this is his favourite part of the codebase.

**Comment by mravn-google** [comment]
Superseded by #11924.

**Pull request closed without merging**